samples/use-onchip-memory/use-onchip-memory.cpp

```diff
@@ -0,0 +1,142 @@
+ *
+ **************************************************************************/
+
+#include <CL/sycl.hpp>
+#include <SYCL/codeplay.hpp>
+#include <iostream>
+
+namespace sycl = cl::sycl;
+namespace access = sycl::access;
+
+// Sample kernel
+// -------------
+// Since kernels are no longer allowed to be declared inline, I dislike
+// declaring a random name in the global namespace that can potentially be
+// repurposed.
+//
+// This kernel performs the same operation as std::iota, but also scales
+// the result by two.
+//
+class scaled_iota {
+ public:
+  using accessor = sycl::accessor<int, 1, access::mode::discard_write,
+                                  access::target::global_buffer>;
+
+  explicit scaled_iota(accessor deviceAccess) noexcept
+      : m_deviceAccess{std::move(deviceAccess)} {}
+
+  void operator()(sycl::nd_item<2> itemId) const noexcept {
+    const auto linearId = itemId.get_global_linear_id();
+    m_deviceAccess[linearId] = linearId * 2;
+  }
+
+ private:
+  accessor m_deviceAccess;
+};
+
+namespace codeplay = sycl::codeplay;
+
+template <class Policy>
+void use_with_policy(Policy policy, sycl::queue& queue) {
+  auto hostData = sycl::vector_class<int>(1024);
+  {
+    auto taskContext = queue.get_context();
+
+    // Notice that the Codeplay property takes a policy argument: this is used
+    // to indicate whether the property is advantageous or genuinely necessary.
+    //
+    auto deviceData = sycl::buffer<int, 1>{
+        hostData.data(),                  //
+        sycl::range<1>(hostData.size()),  //
+        sycl::property_list{
+            sycl::property::buffer::context_bound(taskContext),
+            codeplay::property::buffer::use_onchip_memory(policy)  // <--------
+        }  //                                            ^---------------------
+    };
+
+    deviceData.set_final_data(hostData.data());
+
+    queue.submit([&](sycl::handler & cgh) {
+      constexpr auto dimension_size = 2;
+      auto r = sycl::nd_range<dimension_size>{
+          sycl::range<dimension_size>{
+              hostData.size() / dimension_size,  //
+              dimension_size                     //
+          },
+          sycl::range<dimension_size>{dimension_size, 1}  //
+      };
+      cgh.parallel_for(
+          r,
+          scaled_iota(deviceData.get_access<access::mode::discard_write>(cgh)));
+    });
+
+    queue.wait_and_throw();
+  }
+}
+
+// Codeplay policy extensions have two different enabling mechanisms: the first
+// is to indicate that a policy is preferred. Using this policy means that if
+// the system supports the feature, then the feature will be enabled. If the
+// feature is not present on the system, then it will not be enabled.
+//
+// Puns aside, this is the preferred default.
+//
+void how_to_use_with_prefer() {
+  auto queue = sycl::queue{};
+  ::use_with_policy(codeplay::property::prefer, queue);
+}
+
+// Alternatively, if you can guarantee that your system will support this
```

Comment:
You could also mention here that making it required lets your program know if it is available or not. Preferred will choose one but there's no way to know which. (I don't know if that's useful, maybe you tweak the algorithm slightly if onchip is available or something)

Reply:
Isn't that what I've said?

Reply:
Yes, sort of. I guess the consequence flows naturally, and is a slightly more unusual use-case.